test/kubernetes_unittest.cc

```diff
@@ -1215,4 +1184,705 @@ TEST_F(KubernetesTestFakeServer, MetadataQuery) {
   EXPECT_EQ(pod_metadata->ToString(), m[3].metadata().metadata->ToString());
 }
 
+class KubernetesTestFakeServerOneWatchRetryNodeLevelMetadata
+  : public KubernetesTest {
+ protected:
+  void SetUp() override {
+    server.reset(new testing::FakeServer());
+    KubernetesTest::SetUp();
+  }
+
+  std::unique_ptr<Configuration> CreateConfig() override {
+    return std::unique_ptr<Configuration>(
+      new Configuration(std::istringstream(
+        "InstanceId: TestID\n"
+        "InstanceResourceType: gce_instance\n"
+        "InstanceZone: TestZone\n"
+        "KubernetesClusterLevelMetadata: false\n"  // node-level
+        "KubernetesClusterLocation: TestClusterLocation\n"
+        "KubernetesClusterName: TestClusterName\n"
+        "KubernetesEndpointHost: " + server->GetUrl() + "\n"
+        "KubernetesNodeName: TestNodeName\n"
+        "MetadataIngestionRawContentVersion: TestVersion\n"
+        "KubernetesUpdaterWatchConnectionRetries: 1\n"
+        "KubernetesUseWatch: true\n"
+      )));
+  }
+
+  std::unique_ptr<testing::FakeServer> server;
+};
+
+class KubernetesTestFakeServerOneWatchRetryClusterLevelMetadata
+  : public KubernetesTest {
+ protected:
+  void SetUp() override {
+    server.reset(new testing::FakeServer());
+    KubernetesTest::SetUp();
+  }
+
+  std::unique_ptr<Configuration> CreateConfig() override {
+    return std::unique_ptr<Configuration>(
+      new Configuration(std::istringstream(
+        "InstanceId: TestID\n"
+        "InstanceResourceType: gce_instance\n"
+        "InstanceZone: TestZone\n"
+        "KubernetesClusterLevelMetadata: true\n"  // cluster-level
+        "KubernetesClusterLocation: TestClusterLocation\n"
+        "KubernetesClusterName: TestClusterName\n"
+        "KubernetesEndpointHost: " + server->GetUrl() + "\n"
+        "KubernetesNodeName: TestNodeName\n"
+        "MetadataIngestionRawContentVersion: TestVersion\n"
+        "KubernetesUpdaterWatchConnectionRetries: 1\n"
+        "KubernetesUseWatch: true\n"
+      )));
+  }
+
+  std::unique_ptr<testing::FakeServer> server;
+};
+
+// Polls store until collected_at for resource is newer than
+// last_timestamp.  Returns false if newer timestamp not found after 3
+// seconds (polling every 100 millis).
+bool WaitForNewerCollectionTimestamp(const MetadataStore& store,
```

Comment:
These helper functions should be in an anonymous namespace...

Reply:
Done.